File: wopr-audit/nodejs/src/hooks/webhook.js
const SEVERITY_ORDER = { info: 0, warning: 1, error: 2, critical: 3 };

export class WebhookHook {
  constructor(url, minSeverity = "warning") {
    this.url = url;
    this.minSeverity = minSeverity;
  }

  async process(event) {
    if ((SEVERITY_ORDER[event.severity] || 0) < (SEVERITY_ORDER[this.minSeverity] || 0)) {
      return;
    }
    try {
      const msg = `[${event.severity.toUpperCase()}] ${event.service}: ${event.action}`;
      await fetch(this.url, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          service: event.service,
          severity: event.severity,
          action: event.action,
          message: msg,
          timestamp: event.timestamp,
          correlation_id: event.correlation_id,
        }),
      });
    } catch {
      // webhook failure should never crash the app
    }
  }
}

File: wopr-audit/python/src/wopr_audit/hooks/webhook.py
from wopr_audit.schema import AuditEvent, Severity


class WebhookHook:
    def __init__(self, url: str, min_severity: Severity = Severity.WARNING):
        self.url = url
        self.min_severity = min_severity
        self._severity_order = {
            Severity.INFO: 0, Severity.WARNING: 1,
            Severity.ERROR: 2, Severity.CRITICAL: 3,
        }

    async def process(self, event: AuditEvent) -> None:
        if self._severity_order.get(event.severity, 0) < self._severity_order.get(self.min_severity, 0):
            return
        try:
            import httpx
            async with httpx.AsyncClient(timeout=5.0) as client:
                await client.post(self.url, json={
                    "service": event.service,
                    "severity": event.severity.value,
                    "action": event.action,
                    "message": f"[{event.severity.value.upper()}] {event.service}: {event.action}",
                    "timestamp": event.timestamp.isoformat(),
                    "correlation_id": event.correlation_id,
                    "metadata": event.metadata,
                })
        except Exception:
            pass  # webhook failure should never crash the app
